### PR TITLE
feat: support hostpath local PV

### PR DESCRIPTION
**Pull request by zjx20**
This PR ports the hostpath LocalPV provisioner of OpenEBS to this project, currently only basic provisioning is supported.

Usage:

* Install open-local (this is NOT working now because the docker image hasn’t been updated)

```bash
helm install open-local ./helm
```

* Create test PVC (with the open-local-hostpath StorageClass) and Pod

```bash
kubectl apply -f - <<-"EOF"
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: test-localpv-pvc
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 5Gi
  storageClassName: open-local-hostpath
---
apiVersion: v1
kind: Pod
metadata:
  name: test-localpv
spec:
  restartPolicy: Never
  containers:
   - name: test
     image: bash:latest
     command:
     - bash
     - -c
     - |
       echo `hostname` >> /data/hello.txt
       cat /data/hello.txt
     volumeMounts:
       - mountPath: /data
         name: dir
  volumes:
   - name: dir
     persistentVolumeClaim:
       claimName: test-localpv-pvc
EOF
```

TODO:
* Support project quota and cgroup based IO throttling
* Change the driver name to apecloud.com
* Migrate provisioner-hostpath to CSI based implementation
* Integrate with scheduler

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,11 +1,37 @@
-FROM golang:1.18 AS builder
+FROM --platform=${BUILDPLATFORM} golang:1.21 AS builder
 
-WORKDIR /go/src/github.com/apecloud/open-local
-COPY . .
```
Comment (cnut) [inline Dockerfile:4]: which is best practises (1. prebuild and just copy, 2. build from source code then copy)?

File: cmd/provisioner/provisioner.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+	}
+
+	kubeClient, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return errors.Wrap(err, "unable to get k8s client")
+	}
+
+	//Create a context to receive shutdown signal to help
+	// with graceful exit of the provisioner.
+	ctx := context.Background()
```
Comment (cnut) [inline cmd/provisioner/provisioner.go:80]: openlocal use signal to exit the background process

File: helm/templates/storage-class.yaml [hunk 3]
```diff
@@ -57,7 +57,7 @@ kind: StorageClass
 metadata:
   name: {{ .Values.storageclass.lvm_throttle.name }}
 {{ include "local.labels" . | indent 2 }}
-provisioner: local.csi.aliyun.com
+provisioner: {{ .Values.driver }}
```
Comment (cnut) [inline helm/templates/storage-class.yaml:60]: is this temp solution to work around the provisioner conflict with alibaba-openlocal?

File: pkg/provisioner/env.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
+/*
+Copyright (c) 2023 ApeCloud, Inc. All rights reserved.
```
Comment (cnut) [inline pkg/provisioner/env.go:2]: whether we need to retain the original copyright from openebs, then add PORT copyright

Comment (cnut) [reply pkg/provisioner/env.go:2]:
```
//
// Portions Copyright (c) 2023 ApeCloud, Inc.
//
// Copyright 2006, Google Inc.
// All rights reserved.
```

File: cmd/provisioner/options.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+*/
+
+package provisioner
+
+import (
+	"github.com/spf13/pflag"
+)
+
+type provisionerOptions struct {
+	Master      string
```
Comment (cnut) [inline cmd/provisioner/options.go:24]: master is too general, maybe we can use a better concrete name.

File: pkg/provisioner/provisioner.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+	}
+	p.getVolumeConfig = p.GetVolumeConfig
+
+	return p, nil
+}
+
+// SupportsBlock will be used by controller to determine if block mode is
+//
+//	supported by the host path provisioner.
+func (p *Provisioner) SupportsBlock(_ context.Context) bool {
```
Comment (cnut) [inline pkg/provisioner/provisioner.go:60]: this code is useless?

File: Dockerfile [hunk 7]
```diff
@@ -1,11 +1,37 @@
...
+    --mount=type=cache,target=/go/pkg/mod \
+    go mod download -x
+
+RUN --mount=type=bind,target=. \
+    --mount=type=cache,target=/root/.cache/go-build \
+    --mount=type=cache,target=/go/pkg/mod \
+    go env && \
+    GOOS=${TARGETOS} GOARCH=${TARGETARCH} OUTPUT_DIR=/out make build
+
+FROM alpine:3.18
```
Comment (JingchengLi) [inline Dockerfile:32]: use this image is not recommended?

File: cmd/provisioner/provisioner.go (hunk 2 above)
Comment (zjx20) [reply cmd/provisioner/provisioner.go]: Fixed.

File: pkg/provisioner/env.go (hunk 4 above)
Comment (zjx20) [reply pkg/provisioner/env.go]: Updated.

File: Dockerfile (hunk 7 above)
Comment (JingchengLi) [reply Dockerfile:32]: in sfs csi, I used registry.access.redhat.com/ubi8/ubi:8.8

Comment (zjx20) [reply Dockerfile:32]: Let's do this in other PR. ref #10

File: pkg/provisioner/provisioner_hostpath.go [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+		selectedNodeTaints: taints,
+		imagePullSecrets:   imagePullSecrets,
+	}
+	iErr := p.createInitPod(ctx, podOpts)
+	if iErr != nil {
+		klog.Errorf("Initialize volume %v failed: %v", name, iErr)
+		return nil, pvController.ProvisioningFinished, iErr
+	}
+
+	// TODO(x.zhou): set the usage quota
```
Comment (cnut) [inline pkg/provisioner/provisioner_hostpath.go:86]: is there any drawbacks in openebs quota setting?

**Review by cnut — APPROVED** [review]

File: pkg/provisioner/provisioner_hostpath.go [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+		klog.Errorf("Failed to provision Local PV %s, unable to get path from volume config, err: %v", opts.PVName, err)
+		return nil, pvController.ProvisioningFinished, err
+	}
+
+	imagePullSecrets := GetImagePullSecrets(getImagePullSecrets())
+
+	klog.Infof("Creating volume %v at node with labels {%v}, path:%v,ImagePullSecrets:%v", name, nodeAffinityLabels, path, imagePullSecrets)
+
+	//Before using the path for local PV, make sure it is created.
+	initCmdsForPath := []string{"mkdir", "-m", "0777", "-p"}
```
Comment (gufeijun) [inline pkg/provisioner/provisioner_hostpath.go:70]: move this to method 'createInitPod'?

**Pull request merged**